### PR TITLE
snapshot: feedback multi rocksdb

### LINKED ISSUE #14436: Record the duration of executing the snapshot in raftstore-v2
## Development Task
ref issue: https://github.com/tikv/pd/issues/5467
In this issue, PD will use the snapshot duration to control the current size of the move-peer

ref #13409

### PR DESCRIPTION
**Pull request by bufferflies**
### What is changed and how it works?

Issue Number: Close #14436

What's Changed:

```commit-message
```

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/server/tablet_snap.rs [hunk 1]
```diff
@@ -634,6 +638,17 @@ pub fn send_snap(
         msg.get_message().get_snapshot(),
     );
 
+    let (snap_start, generate_duration_sec) = {
+        let snap = msg.get_message().get_snapshot();
+        let mut snap_data = RaftSnapshotData::default();
+        if let Err(e) = snap_data.merge_from_bytes(snap.get_data()) {
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply src/server/tablet_snap.rs:644]: yes, I will use lock to share these message.

File: components/raftstore/src/store/async_io/read.rs [hunk 2]
```diff
@@ -246,6 +243,13 @@ where
                         "total_size" => total_size,
                         "total_keys" => total_keys,
                     );
+                    self.snap_mgr()
+                        .begin_snapshot(snap_key, start, generate_duration_secs);
+                    snapshot.set_data(snap_data.write_to_bytes().unwrap().into());
```
Comment (pingyu) [inline components/raftstore/src/store/async_io/read.rs:248]: Duplicated with line 225.

File: components/raftstore/src/store/snap.rs [hunk 3]
```diff
@@ -1994,9 +1995,49 @@ impl TabletSnapManager {
...
+            });
+    }
+
+    pub fn stats(&self) -> SnapStats {
+        let finished: HashMap<TabletSnapKey, (Instant, SnapshotStat)> = self
+            .stats
+            .lock()
+            .unwrap()
+            .drain_filter(|_, (_, stat)| stat.get_region_id() > 0)
+            .collect();
```
Comment (pingyu) [inline components/raftstore/src/store/snap.rs:2027]: The `collect` may be not necessary. `.drain_filter` return an iterator, and can be chained to following `map` and `filter`.

File: src/server/tablet_snap.rs [hunk 4]
```diff
@@ -81,7 +82,11 @@ fn is_sst(file_name: &str) -> bool {
 
 async fn read_to(f: &mut File, to: &mut Vec<u8>, size: usize, limiter: &Limiter) -> Result<()> {
     // It's likely in page cache already.
-    limiter.consume(size / 2).await;
+    let cost = size / 2;
+    limiter.consume(cost).await;
+    SNAP_LIMIT_TRANSPORT_BYTES_COUNTER_STATIC
+        .send
+        .inc_by(cost.try_into().unwrap());
```
Comment (pingyu) [inline src/server/tablet_snap.rs:89]: Try `cost as u64`.

File: src/server/tablet_snap.rs [hunk 5]
```diff
@@ -412,6 +420,9 @@ async fn recv_snap_files<'a>(
         (0, vec![])
     };
     let received = accept_missing(&path, missing_ssts, &mut stream, &limiter).await?;
+    SNAP_LIMIT_TRANSPORT_BYTES_COUNTER_STATIC
+        .recv
+        .inc_by(received);
```
Comment (pingyu) [inline src/server/tablet_snap.rs:425]: Ditto.

File: components/raftstore/src/store/snap.rs [hunk 6]
```diff
@@ -1994,9 +1995,49 @@ impl TabletSnapManager {
         Ok(Self {
             base: path,
             receiving: Arc::default(),
+            stats: Arc::default(),
         })
     }
 
+    pub fn begin_snapshot(&self, key: TabletSnapKey, start: Instant, generate_duration_sec: u64) {
```
Comment (pingyu) [inline components/raftstore/src/store/snap.rs:2002]: Suggest to add an unit test for these new methods.

File: components/raftstore/src/store/async_io/read.rs (hunk 2 above)
Comment (bufferflies) [reply components/raftstore/src/store/async_io/read.rs]: fixed

File: src/server/tablet_snap.rs (hunk 4 above)
Comment (bufferflies) [reply src/server/tablet_snap.rs]: fixed

File: src/server/tablet_snap.rs (hunk 5 above)
Comment (bufferflies) [reply src/server/tablet_snap.rs]: fixed

File: components/raftstore/src/store/snap.rs (hunk 3 above)
Comment (bufferflies) [reply components/raftstore/src/store/snap.rs]: fixed

File: components/raftstore/src/store/snap.rs (hunk 6 above)
Comment (bufferflies) [reply components/raftstore/src/store/snap.rs:2002]: already add unit test for it

**Review by pingyu — APPROVED** [review]
LGTM~

**Review by lhy1024 — APPROVED** [review]

**Pull request merged**